cdap-watchdog/src/main/java/co/cask/cdap/logging/logbuffer/LogBufferService.java

```diff
@@ -85,6 +89,10 @@ protected void startUp() throws Exception {
     // start all the log pipelines
     validateAllFutures(Iterables.transform(pipelines, Service::start));
 
+    // start log recovery service to recover all the pending logs
+    recoveryService = new LogBufferRecoveryService(cConf, bufferPipelines, checkpointManagers);
+    recoveryService.startAndWait();
```

Comment:
Is it okay for the recovery service to be running when other services start processing files? Won't the recovery service also start processing the same files again?

Reply:
While recovery service is running, other services should not start log processing. To take care of that case, we use  same thread executor in `LogBufferRecoveryService`. Added comment here to clarify that.

Reply:
However, with this approach startUp method of LogBufferService will be blocked until recovery happens.

Reply:
Discussed this with @chtyim . If recovery service is taking too long, other services, wont be able to publish logs. Because of that, we would first start the recovery service and then start http log service. The trade off would be, if there is heavy inflow of logs through http service, recovered logs can be out of order.